benchmarking/daint_single_node/stdout_to_json.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+
+# This script parses the standard output of a FV3GFS run to extract timing
+# information and convert it to the JSON format which is used by the performance
+# monitoring script.
+
+# 2021/01/22 Oliver Fuhrer, Vulcan Inc, oliverf@vulcan.com
+
+import os, re, click, glob, time, datetime, json, sys
+
+# this dict maps the names of the timer in the JSON file to the actual
+# timers reported in stdout of FV3GFS, multiple entries are accumulated
+TIMER_MAPPING = {
+  "total": ["Total runtime"],
+  "init": ["1-Initialization", "2-Main-loop-1st-trip"],
+  "main_loop": ["3-Main-loop"],
+  "fv_dynamics": ["3.1.1-fv_dynamics"],
+  "dyn_core": ["3.1.1.1-dyn_core"],
+  "tracer_adv": ["3.1.1.2-Tracer-advection"],
+  "remapping": ["3.1.1.3-Remapping"]
+}
+
+def num(s):
+    try:
+        return int(s)
+    except ValueError:
+        return float(s)
+
+
+@click.command()
+@click.option('--stdout_file_regex', default='slurm-*.out', help='regex pattern to identify stdout file')
+@click.argument('run_directory', nargs=1)
+def stdout_to_json(stdout_file_regex, run_directory):
+
+    # find stdout file
+    files = glob.glob(os.path.join(run_directory, stdout_file_regex))
+    assert len(files) == 1, 'Either no or too many matches for given stdout file name regex pattern'
+    stdout_file = files[0]
+
+    # extract timer string from stdout
+    with open(stdout_file, 'r+') as f:
+        stdout = f.read()
+    match = re.search(r'^Total runtime.*^4-Termination', stdout, re.MULTILINE | re.DOTALL)
+    assert match, 'Issue extracting timings from stdout of SLURM job'
+    
+    # parse raw timers
+    raw_timers = {}
+    labels = ['hits', 'tmin', 'tmax', 'tavg', 'tstd', 'tfrac', 'grain', 'pemin', 'pemax']
+    for line in match.group().splitlines():
+        name = line[0:32].strip()
+        values = [num(val) for val in line[32:].split()]
+        raw_timers[name] = dict(zip(labels, values))
+
+    # convert into format for plotting
+    times = {}
+    for json_name, fv3_names in TIMER_MAPPING.items():
+        times[json_name] = {"hits": 0, "minimum": 0., "maximum": 0., "median": 0.}
```

Comment:
this should not be necessary

Reply:
Since some of the JSON timers are sums of multiple timers (e.g. "init"), I've implement this strategy where we initialize to zero and then accumulate. Make sense?

Reply:
ah got it. Makes sense